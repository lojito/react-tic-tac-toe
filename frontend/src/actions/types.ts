import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
