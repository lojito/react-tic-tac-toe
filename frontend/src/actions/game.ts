import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

import {
  CategoryAction,
  ComputerImageAction,
  DisabledAction,
  FirstAction,
  GameOverAction,
  PlayingLevelAction,
  ResultAction,
  UserImageAction,
  WinnersSquaresAction,
} from "./types";

export const CHANGE_CATEGORY = "CHANGE_CATEGORY";
export const CHANGE_DISABLED = "CHANGE_DISABLED";
export const CHANGE_FIRST = "CHANGE_FIRST";
export const CHANGE_IMAGES = "CHANGE_IMAGES";
export const CHANGE_USER_IMAGE = "CHANGE_USER_IMAGE";
export const CHANGE_COMPUTER_IMAGE = "CHANGE_COMPUTER_IMAGE";
export const CHANGE_PLAYING_LEVEL = "CHANGE_PLAYING_LEVEL";
export const CHANGE_GAME_OVER = "CHANGE_GAME_OVER";
export const CHANGE_GAME_RESULT = "CHANGE_GAME_RESULT";
export const CHANGE_WINNERS_SQUARES = "CHANGE_WINNERS_SQUARES";

export const changeCategory = (category: Categories): CategoryAction => {
  return {
    type: CHANGE_CATEGORY,
    category,
  };
};

export const changeDisabled = (disabled: boolean): DisabledAction => {
  return {
    type: CHANGE_DISABLED,
    disabled,
  };
};

export const changeFirst = (first: FirstPlayer): FirstAction => {
  return {
    type: CHANGE_FIRST,
    first,
  };
};

export const changeUserImage = (userImage: ImageName): UserImageAction => {
  return {
    type: CHANGE_USER_IMAGE,
    userImage,
  };
};

export const changeComputerImage = (
  computerImage: ImageName
): ComputerImageAction => {
  return {
    type: CHANGE_COMPUTER_IMAGE,
    computerImage,
  };
};

export const changeLevel = (level: PlayingLevel): PlayingLevelAction => {
  return {
    type: CHANGE_PLAYING_LEVEL,
    level,
  };
};

export const changeOver = (over: boolean): GameOverAction => {
  return {
    type: CHANGE_GAME_OVER,
    over,
  };
};

export const changeResult = (result: GameResult): ResultAction => {
  return {
    type: CHANGE_GAME_RESULT,
    result,
  };
};

export const changeWinners = (
  winners: SquareLocation[]
): WinnersSquaresAction => {
  return {
    type: CHANGE_WINNERS_SQUARES,
    winners,
  };
};
