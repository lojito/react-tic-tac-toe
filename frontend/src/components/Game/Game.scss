.game-wrapper {
  margin-top: 10px;
  text-align: center;
}

.buttons {
  margin: 5px 0;

  .see-game, .back {
    margin: 0 5px;

    a {
      color: black;
    }
  }

  .delete-game {
    margin: 0 5px;
    cursor: pointer;
   }

  .back {
    cursor: pointer;
  }
}

hr {
  margin: 15px;
}