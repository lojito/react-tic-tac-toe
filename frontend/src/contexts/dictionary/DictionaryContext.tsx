import { createContext } from "react";
import { Language } from "../../contexts/language/LanguageContext";
import data from "../../data/json/dictionary.json";

export type DictionaryEntry = Partial<
  Record<
    | "ABOUT"
    | "ABOUT_GAME_DEVELOPED_WITH"
    | "ABOUT_OTHER_HOME_PROJECTS"
    | "ABOUT_SOURCE_CODE"
    | "AUTH_FORM_EMAIL_LABEL"
    | "AUTH_FORM_EMAIL_PLACEHOLDER"
    | "AUTH_FORM_NAME_LABEL"
    | "AUTH_FORM_NAME_PLACEHOLDER"
    | "AUTH_FORM_PASSWORD_LABEL"
    | "AUTH_FORM_PASSWORD_PLACEHOLDER"
    | "AUTH_ERROR_SERVER_NOT_RESPONDING"
    | "AUTH_ERROR_VALIDATION_FAILED"
    | "AUTH_ERROR_UNKNOWN"
    | "BUTTON_BACK"
    | "BUTTON_DELETE_GAME"
    | "BUTTON_LOGOUT"
    | "BUTTON_PLAY"
    | "BUTTON_PLAY_AGAIN"
    | "BUTTON_SEE_GAME"
    | "BUTTON_SIGNIN"
    | "BUTTON_SIGNUP"
    | "CATEGORY"
    | "CATEGORY_ANIMALS"
    | "CATEGORY_FRUITS_AND_VEGETABLES"
    | "CATEGORY_GERMANY_LANDMARKS"
    | "CATEGORY_HAVANA_LANDMARKS"
    | "CATEGORY_MONTREAL_LANDMARKS"
    | "CATEGORY_PUPPIES"
    | "CATEGORY_SEINFELD"
    | "CATEGORY_SOCCER_PLAYERS"
    | "CATEGORY_SPAIN_LANDMARKS"
    | "CATEGORY_VANCOUVER_LANDMARKS"
    | "CONTACT"
    | "CONTACT_AUTHOR"
    | "CONTACT_EMAIL"
    | "CONTACT_GITHUB"
    | "CONTACT_LINKEDIN"
    | "FLAG_ENGLISH_HINT"
    | "FLAG_FRENCH_HINT"
    | "FLAG_SPANISH_HINT"
    | "GAME"
    | "GAME_OVER"
    | "GAME_STARTED_BY"
    | "GAME_STARTED_BY_COMPUTER"
    | "GAME_STARTED_BY_YOU"
    | "GAME_DELETING"
    | "GAME_DELETE_ERROR"
    | "GAME_SAVE_ERROR"
    | "GAME_SAVING"
    | "HISTORICAL"
    | "HISTORICAL_NO_DATA_YET"
    | "HISTORICAL_FETCHING_GAMES"
    | "HISTORICAL_FETCHING_GAME"
    | "HISTORICAL_FETCHING_GAMES_ERROR"
    | "HISTORICAL_FETCHING_GAME_ERROR"
    | "IMAGE_COMPUTER_LABEL"
    | "IMAGE_COMPUTER_TOOLTIP"
    | "IMAGE_REFRESH_TOOLTIP"
    | "IMAGE_USER_LABEL"
    | "IMAGE_USER_TOOLTIP"
    | "LOGIN"
    | "MESSAGE_DRAW"
    | "MESSAGE_LOST"
    | "MESSAGE_WON"
    | "PLAYING_LEVEL_EASY"
    | "PLAYING_LEVEL_LABEL"
    | "PLAYING_LEVEL_SMART"
    | "PLAYING_LEVEL_TOOLTIP"
    | "REPOSITORY"
    | "SIGNUP"
    | "WELCOME_BACK"
    | "START_GAME_LABEL"
    | "START_GAME_ANSWER_NO"
    | "START_GAME_TOOLTIP"
    | "START_GAME_ANSWER_YES",
    string
  >
>;

export type DictionaryContextValue = {
  dictionary: DictionaryEntry;
};

export const DictionaryContext = createContext<DictionaryContextValue>({
  dictionary: data[Language.English],
});
