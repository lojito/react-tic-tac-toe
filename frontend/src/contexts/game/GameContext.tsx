import React, { createContext } from "react";
import { Action } from "../../actions/types";
import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../../types";

export type State = {
  category: Categories;
  userImage: ImageName;
  computerImage: ImageName;
  first: FirstPlayer;
  level: PlayingLevel;
  disabled: boolean;
  over: boolean;
  result: GameResult;
  winners: SquareLocation[];
};

export const GameContext = createContext<{
  game: State;
  dispatch: React.Dispatch<Action>;
}>({ game: {} as State, dispatch: {} as React.Dispatch<Action> });
