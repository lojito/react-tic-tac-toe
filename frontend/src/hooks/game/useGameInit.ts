import { useCallback } from "react";
import { changeDisabled, changeOver, changeResult } from "../../actions/game";
import { Action } from "../../actions/types";
import { GameResult } from "../../types";

function useGameInit(dispatch: React.Dispatch<Action>) {
  const handleGameInit = useCallback(() => {
    dispatch(changeOver(false));
    dispatch(changeDisabled(false));
    dispatch(changeResult(GameResult.Started));
  }, [dispatch]);

  return { handleGameInit };
}

export default useGameInit;
