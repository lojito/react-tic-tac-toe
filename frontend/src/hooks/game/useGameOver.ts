import { useCallback } from "react";
import { changeOver, changeResult, changeWinners } from "../../actions/game";
import { Action } from "../../actions/types";
import { DBGame } from "../../types";

function useGameOver(dispatch: React.Dispatch<Action>) {
  const handleGameOver = useCallback(
    (game: DBGame) => {
      dispatch(changeResult(game.result));
      dispatch(changeOver(true));
      dispatch(changeWinners(game.winners));
    },
    [dispatch]
  );

  return { handleGameOver };
}

export default useGameOver;
