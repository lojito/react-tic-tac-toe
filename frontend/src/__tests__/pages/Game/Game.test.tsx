import React from "react";
import { useMutation } from "@apollo/client";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DictionaryContext } from "../../../contexts/dictionary/DictionaryContext";
import { GameContext, State } from "../../../contexts/game/GameContext";
import useGameInit from "../../../hooks/game/useGameInit";
import useGameOver from "../../../hooks/game/useGameOver";
import useGameStart from "../../../hooks/game/useGameStart";
import Game from "../../../pages/Game/Game";
import { GameResult, Player, PlayingLevel } from "../../../types";

jest.mock("../../../components/Category/Category", () => () => (
  <div>Category: Soccer players</div>
));

jest.mock("../../../components/Image/Image", () => () => (
  <>
    <div>User: Messi</div>
    <div>Computer: Ronaldo</div>
  </>
));

jest.mock("../../../components/StartGame/StartGame", () => () => (
  <div>You start: Yes</div>
));

jest.mock("../../../components/PlayingLevel/PlayingLevel", () => () => (
  <div>Playing Level: Easy</div>
));

jest.mock(
  "../../../components/BoardUI/BoardUI",
  () =>
    ({ onGameOver }: { onGameOver: () => void }) => {
      return (
        <button aria-label="board" onClick={() => onGameOver()}>
          BoardUI
        </button>
      );
    }
);

jest.mock("../../../hooks/game/useGameStart");
jest.mock("../../../hooks/game/useGameOver");
jest.mock("../../../hooks/game/useGameInit");

jest.mock("@apollo/client", () => ({
  useMutation: jest.fn(),
}));

describe("Game", () => {
  const mockHandleGameOver = jest.fn();

  const mockDictionary = {
    BUTTON_PLAY: "Start the game",
    BUTTON_PLAY_AGAIN: "Play again",
    CATEGORY_ANIMALS: "Animals",
    CATEGORY_FRUITS_AND_VEGETABLES: "Fruits & Vegetables",
    CATEGORY_GERMANY_LANDMARKS: "Germany landmarks",
    CATEGORY_HAVANA_LANDMARKS: "Havana landmarks",
    CATEGORY_MONTREAL_LANDMARKS: "Montreal landmarks",
    CATEGORY_PUPPIES: "Puppies",
    CATEGORY_SEINFELD: "Seinfeld",
    CATEGORY_SOCCER_PLAYERS: "Soccer players",
    CATEGORY_SPAIN_LANDMARKS: "Spain landmarks",
    CATEGORY_VANCOUVER_LANDMARKS: "Vancouver landmarks",
    CATEGORY: "Category",
    GAME_SAVE_ERROR:
      "An error ocurred while saving the game results into the database",
    GAME_SAVING: "Saving the game data...",
    MESSAGE_WON: "You won!",
    MESSAGE_LOST: "You lost!",
    MESSAGE_DRAW: "It's a draw!",
    WELCOME_BACK: "Welcome back",
  };

  const mockAddGame = jest.fn();

  interface GameWithProviderProps {
    result?: GameResult;
    gameOver?: boolean;
    disabled?: boolean;
    addGame?: jest.Mock;
    name?: string;
  }

  const TestComponent = ({
    disabled = false,
    gameOver = false,
    result = GameResult.Started,
    name = "Livan",
  }: GameWithProviderProps = {}) => {
    const game = {
      category: {
        id: 0,
        folder: "habana",
        name: "CATEGORY_HAVANA_LANDMARKS",
      },
      userImage: 0,
      computerImage: 1,
      first: Player.User,
      level: PlayingLevel.Easy,
    } as State;
    game.disabled = disabled;
    game.over = gameOver;
    game.result = result;

    const dispatch = jest.fn();

    return (
      <DictionaryContext.Provider value={{ dictionary: mockDictionary }}>
        <GameContext.Provider value={{ game, dispatch }}>
          <Game token="token" name={name} />
        </GameContext.Provider>
      </DictionaryContext.Provider>
    );
  };

  const renderGame = ({
    result = GameResult.Started,
    gameOver = false,
    disabled = false,
    addGame = mockAddGame,
  }: GameWithProviderProps = {}) => {
    (useGameInit as jest.Mock).mockReturnValue({
      handleGameInit: jest.fn,
    });

    (useGameOver as jest.Mock).mockReturnValue({
      handleGameOver: mockHandleGameOver,
    });

    (useGameStart as jest.Mock).mockReturnValue({
      handleGameStart: jest.fn,
    });

    (useMutation as jest.Mock).mockReturnValue([
      addGame,
      { loading: false, error: "" },
    ]);

    const name = "Livan";
    const utils = render(
      <TestComponent
        disabled={disabled}
        name={name}
        gameOver={gameOver}
        result={result}
      />
    );

    const welcomeBack = screen.getByText(`Welcome back, ${name}`);

    const categorySoccer = screen.getByText(/Category: soccer players/i);
    const userImage = screen.getByText(/User: Messi/i);
    const computerImage = screen.getByText(/Computer: Ronaldo/i);
    const startGameYou = screen.getByText(/You start: Yes/i);
    const playingLevelEasy = screen.getByText(/Playing Level: Easy/i);

    let message: HTMLElement | null;
    if (result === GameResult.Lost) {
      message = screen.getByText(mockDictionary.MESSAGE_LOST);
    } else if (result === GameResult.Draw) {
      message = screen.getByText(mockDictionary.MESSAGE_DRAW);
    } else if (result === GameResult.Won) {
      message = screen.getByText(mockDictionary.MESSAGE_WON);
    } else {
      message = null;
    }

    const buttonPlay = screen.getByRole("button", { name: "play" });
    const buttonBoard = screen.getByRole("button", { name: "board" });

    return {
      ...utils,
      mockDictionary,
      welcomeBack,
      categorySoccer,
      userImage,
      computerImage,
      startGameYou,
      playingLevelEasy,
      message,
      buttonPlay,
      buttonBoard,
    };
  };

  it("renders correctly", () => {
    const { asFragment } = renderGame();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the welcome back message", () => {
    const { welcomeBack } = renderGame();

    expect(welcomeBack).toBeInTheDocument();
  });

  it("should display the Category component", () => {
    const { categorySoccer } = renderGame();

    expect(categorySoccer).toBeInTheDocument();
  });

  it("should display the Image component for the user", () => {
    const { userImage } = renderGame();

    expect(userImage).toBeInTheDocument();
  });

  it("should display the Image component for the computer", () => {
    const { computerImage } = renderGame();

    expect(computerImage).toBeInTheDocument();
  });

  it("should display the Start game component", () => {
    const { startGameYou } = renderGame();

    expect(startGameYou).toBeInTheDocument();
  });

  it("should display the Playing level component", () => {
    const { playingLevelEasy } = renderGame();

    expect(playingLevelEasy).toBeInTheDocument();
  });

  it("should display the game lost message", () => {
    const gameOver = true;
    const result = GameResult.Lost;

    const { message: messageLost } = renderGame({ result, gameOver });

    expect(messageLost).toBeInTheDocument();
  });

  it("should display the game won message", () => {
    const gameOver = true;
    const result = GameResult.Draw;

    const { message: messageDraw } = renderGame({ result, gameOver });

    expect(messageDraw).toBeInTheDocument();
  });

  it("should display the game draw message", () => {
    const gameOver = true;
    const result = GameResult.Won;

    const { message: messageWon } = renderGame({ result, gameOver });

    expect(messageWon).toBeInTheDocument();
  });

  it("should not display any message", () => {
    const { mockDictionary } = renderGame();

    expect(
      screen.queryByText(mockDictionary.MESSAGE_LOST)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText(mockDictionary.MESSAGE_DRAW)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText(mockDictionary.MESSAGE_WON)
    ).not.toBeInTheDocument();
  });

  it("should not disable the button before starting the game", () => {
    const { buttonPlay } = renderGame();

    expect(buttonPlay).toBeEnabled();
  });

  it("should not disable the button when game is over", () => {
    const gameOver = true;
    const result = GameResult.Lost;

    const { buttonPlay } = renderGame({ result, gameOver });

    expect(buttonPlay).toBeEnabled();
  });

  it("should disable the button when game has started", () => {
    const gameOver = false;
    const result = GameResult.Started;
    const disabled = true;

    const { buttonPlay } = renderGame({ result, gameOver, disabled });

    expect(buttonPlay).toBeDisabled();
  });

  it("should display the correct text of the button", () => {
    const { buttonPlay, mockDictionary } = renderGame();

    expect(buttonPlay).toHaveTextContent(mockDictionary.BUTTON_PLAY);
  });

  it("should save the game info", async () => {
    const user = userEvent.setup();
    const { buttonBoard } = renderGame();

    await user.click(buttonBoard);

    expect(mockHandleGameOver).toHaveBeenCalled();
  });

  it("should display the loading message while saving the game data", async () => {
    const user = userEvent.setup();
    const { buttonBoard, rerender } = renderGame();

    await user.click(buttonBoard);

    (useMutation as jest.Mock).mockReturnValueOnce([
      mockAddGame,
      { loading: true },
    ]);

    rerender(<TestComponent result={GameResult.Draw} />);

    expect(screen.getByText(mockDictionary.GAME_SAVING)).toBeInTheDocument();
  });

  it("should display an error when the add game handler throws one", async () => {
    const user = userEvent.setup();
    const { buttonBoard, rerender } = renderGame();

    await user.click(buttonBoard);

    (useMutation as jest.Mock).mockReturnValueOnce([
      mockAddGame,
      { error: "There was an error saving the game data." },
    ]);

    rerender(<TestComponent />);

    expect(
      screen.getByText(mockDictionary.GAME_SAVE_ERROR)
    ).toBeInTheDocument();
  });
});
