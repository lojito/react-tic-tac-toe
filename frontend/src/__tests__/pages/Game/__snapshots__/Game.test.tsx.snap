// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly 1`] = `
<DocumentFragment>
  <div
    class="game text-center"
    data-testid="game"
  >
    <div
      class="welcome-back"
    >
      Welcome back, Livan
    </div>
    <div>
      Category: Soccer players
    </div>
    <div>
      User: Messi
    </div>
    <div>
      Computer: Ronaldo
    </div>
    <div
      class="settings"
    >
      <div>
        You start: Yes
      </div>
      <div>
        Playing Level: Easy
      </div>
    </div>
    <button
      aria-label="board"
    >
      BoardUI
    </button>
    <div
      class="board-footer"
    >
      <button
        aria-label="play"
        class="play-again"
      >
        Start the game
      </button>
    </div>
  </div>
</DocumentFragment>
`;
